### PR TITLE
Fix alignment in Go structs

### PR DESCRIPTION
**Pull request by aanm**
There was a structure that was missing a field which was causing a
misalignment in the cilium bpf ct list global.

```
TCP OUT 192.168.33.11 48672:6443 expires=16211 RxPackets=0 RxBytes=0 TxPackets=1 TxBytes=66 Flags=0 RevNAT=3 SourceSecurityID=2038366208
```

This bug triggered a verification of all golang structures to see if
they were missing any field or they were misaligned with the C
equivalent structure.

It doesn't solve the out-of-order fields between the 2 programming
languages as that should be manually verified by the developer.

Now, every time the cilium-agent starts a verification of all C and
Golang equivalent structs are verified to have the exact offset per
field and the exact size between both structures.

In case they mismatch, a similar panic will be presented when running
cilium-agent:

```
different structure types for CGO and GO: align_checker._Ctype_struct_ct_entry != ctmap.CtEntry
panic: C struct field "slave" (2) has different size than GoStruct field "TxFlagsSeen" (1)

goroutine 1 [running]:
github.com/cilium/cilium/pkg/align_checker.init.0()
        /home/aanm/git-repos/go/src/github.com/cilium/cilium/pkg/align_checker/align_checker.go:93 +0x1d8
```

Signed-off-by: André Martins <andre@cilium.io>

```release-note
Fix alignment for go structures to avoid data corruption between BPF maps and Go
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5538)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

File: pkg/maps/ipcache/ipcache.go [hunk 1]
```diff
@@ -56,8 +56,7 @@ const (
 // Must be in sync with struct bpf_ipcache_key in <bpf/lib/eps.h>
 type Key struct {
 	Prefixlen uint32
-	Pad1      uint16
-	Pad2      uint8
+	Pad       [3]uint8
```
Comment (joestringer) [inline pkg/maps/ipcache/ipcache.go]: This was deliberately set this as a uint16 + uint8 because we have had issues in the past with golang arrays during dump which caused panics (https://github.com/cilium/cilium/issues/3491). Let's change the BPF definition instead.

**Review by joestringer — CHANGES_REQUESTED** [review]
Good spotting, shouldn't have any runtime impact just the CLI tools reporting the wrong thing right?

Thanks for taking this on, this should save time in future when changing/extending these structures.

**Comment by aanm** [comment]
@joestringer the verification is done with `cilium-agent` starts, if anything mismatches the agent doesn't even start.

**Comment by joestringer** [comment]
@aanm Sorry, I mean the underlying bug which seems like it should have been in Cilium 1.2 as well (slave field is missing)

**Comment by aanm** [comment]
@joestringer To be honest I'm not sure what's the impact. When doing ct cleanups can't we write more data into the memory where we shouldn't? Because the types mismatch in size.

**Comment by joestringer** [comment]
@aanm Looks like golang doesn't pack structures by default so I suspect that it just shuffles the TxFlagsSeen/RxFlagsSeen and maybe SourceSecurityID fields around, then inserts a 16-bit gap then follows that with LastRxReport starting at the next 64-bit boundary.

Given that we previously saw some pretty serious memory corruption issues with CT entries, it seems likely that this is not causing real issues on master today.

File: pkg/align_checker/align_checker.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		}
+	}
+	if cStruct.Size() != goStruct.Size() {
+		return fmt.Errorf("C struct %q (%d) has different size than GoStruct %q (%d)",
+			cStruct.Name(), cStruct.Size(), goStruct.Name(), goStruct.Size())
+	}
+	return nil
+}
+
+var cToGO = map[reflect.Type]reflect.Type{
```
Comment (joestringer) [inline pkg/align_checker/align_checker.go]: Can we do sizeof() comparisons here too?

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by borkmann** [comment]
One note: great work, big step forward already, I think the 100% safe choice would eventually be to compare the generated BPF object file's BTF data with the struct offset, padding etc from golang. Here, the included header file will still be compiled with (presumably) gcc in the background for the cgo part, so clang/llvm BPF backend could potentially still have slightly different outcome compared to gcc. I would presume it might only be in rare cases though.

**Review by joestringer — APPROVED** [review]

File: bpf/lib/nat46.h [hunk 3]
```diff
@@ -22,6 +22,7 @@
 #include <linux/icmp.h>
 #include <linux/icmpv6.h>
 #include "common.h"
+#include "ipv4.h"
```
Comment (joestringer) [inline bpf/lib/nat46.h:25]: Why is this necessary?

Comment (borkmann) [reply bpf/lib/nat46.h:25]: I've suggested this to @aanm offline this was needed to fix a compile warning since nat46.h uses `ipv4_hdrlen()` which is only available in ipv4.h, and nat46.h gets pulled in via conntrack header.

Comment (aanm) [reply bpf/lib/nat46.h:25]: @joestringer 
```
# command-line-arguments
In file included from bpf/lib/conntrack.h:29,
                 from bpf/main.go:15:
bpf/lib/nat46.h: In function 'ipv4_to_ipv6':
bpf/lib/nat46.h:256:6: warning: implicit declaration of function 'ipv4_hdrlen'; did you mean 'ipv6_hdrlen'? [-Wimplicit-function-declaration]
  if (ipv4_hdrlen(ip4) != sizeof(v4))
      ^~~~~~~~~~~
      ipv6_hdrlen
```

**Comment by joestringer** [comment]
Hit #5512 .

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]
